pkg/pillar/cmd/zedagent/parseconfig.go

```diff
@@ -1421,6 +1421,70 @@ func parseIpspec(ipspec *zconfig.Ipspec,
 		config.DhcpRange.Start = start
 		config.DhcpRange.End = end
 	}
+
+	addrCount := types.GetIPAddrCountOnSubnet(config.Subnet)
+	if addrCount < types.MinSubnetSize {
+		if addrCount == 0 {
+			return nil
+		}
+		return fmt.Errorf("network(%s), has invalid Subnet range",
+			config.Key())
+	}
+
+	// if not set, take some default
+	if config.Gateway == nil {
+		config.Gateway = types.AddToIP(config.Subnet.IP, 1)
+		log.Warnf("network(%s), No Gateway, setting default(%s)",
+			config.Key(), config.Gateway.String())
+	}
+	dhcpRangeStart := 2
+	if addrCount >= types.LargeSubnetSize {
+		// the dhcpRange starts at the half point,
+		// provided the DhcpRange.End is not set
+		if config.DhcpRange.End == nil {
+			dhcpRangeStart = addrCount / 2
+		}
+	}
+	// last addressable endpoint, with 0 base, and subnet.IP as start,
+	// it accounts for (2^(iplen - subnetMask) - 2) addresses
+	dhcpRangeEnd := addrCount - 2
+
+	// if not set, take some default
+	if config.DhcpRange.Start == nil {
+		config.DhcpRange.Start = types.AddToIP(config.Subnet.IP,
+			dhcpRangeStart)
+		log.Warnf("network(%s), No Dhcp Start, setting default(%s)",
+			config.Key(), config.DhcpRange.Start.String())
+	}
+	if config.DhcpRange.End == nil {
+		config.DhcpRange.End = types.AddToIP(config.Subnet.IP,
+			dhcpRangeEnd)
+		log.Warnf("network(%s), No Dhcp End, setting default(%s)",
+			config.Key(), config.DhcpRange.End.String())
+	}
+	// check whether the dhcp range(start, end)
+	// equal (network, gateway, broadcast) addresses
+	if network := types.GetIPNetwork(config.Subnet); network != nil {
+		if network.Equal(config.DhcpRange.Start) {
+			config.DhcpRange.Start =
+				types.AddToIP(config.DhcpRange.Start, 1)
+		}
+		if config.Gateway.Equal(config.DhcpRange.Start) {
+			config.DhcpRange.Start =
```

Comment:
Should we log a warning in this case to tell that we are increasing the DhcpRange Start to avoid collision with the Gateway %s?